### ISSUE TITLE
NixOS Waybar module hyprland/workspaces not found

### ISSUE THREAD

**Issue by steeple-jack** [Alexays/Waybar#2539]
Hi all,

I'm in the middle of a from-scratch NixOS Hyprland build and I'm having issues getting Waybar to work on hyprland. I'm new to NixOS and Hyprland so bare with me :)

Currently, none of the hyprland modules work (workspaces, window, etc)

`module hyprland/workspaces: Unknown module: hyprland/workspaces`

I applied this overlay to Waybar to enable experimental options but this doesn't seem to have worked:

```
  nixpkgs.overlays = [
    (self: super: {
      waybar = super.waybar.overrideAttrs (oldAttrs: {
        mesonFlags = oldAttrs.mesonFlags ++ [ "-Dexperimental=true" ];
      });
    })
  ];
```

I saw some talk of experimental options being enabled by default in the nixpkg but when looking it seems like that hasn't been merged in yet. 

Has anyone else run into this problem and found a solution?

**Comment by jficz** [comment]
Similar issue here, about a week or two ago, waybar stopped displaying workspaces on Hyprland for me on NixOS. I've been using the experimental overlay for months (and it worked previously), no changes made on my side except regular updates.

Weird thing is waybar doesn't show any relevant errors or warnings in console as it did before (when experimental was disabled), that leads me to think this is a different issue. Perhaps in Hyprland since last Waybar release (0.9.22) is from half August.

**Comment by jficz** [comment]
or not, looks like Hyprland support was added to Waybar quite recently (and quietly):
https://github.com/Alexays/Waybar/wiki/Module:-Hyprland

**Comment by zjeffer** [comment]
> Similar issue here, about a week or two ago, waybar stopped displaying workspaces on Hyprland for me on NixOS.

Is this because you were using wlr/workspaces, which stopped working on Hyprland recently? Hyprland/workspaces should be used instead.

---

I know nothing about Nix so I can't really help with the OPs issue, sadly.

**Comment by jficz** [comment]
yes, I wasn't aware that `hyprland/workspaces` even existed until a few hours ago

**Comment by steeple-jack** [comment]
See, in my case, `hyprland/workspaces` doesn't work but `wlr/workspaces` does. This is after I applied the experimental patch. Before, `wlr/workspaces` didn't work either

**Comment by jficz** [comment]
huh, what version of waybar does that overlay build? What nixpkgs are you at? I'm on unstable and I have the overlay in my user's nixpkgs, not systems. And it works for me (`hyprland`, not `wlr`). Using home-manager to config waybar.

**Comment by steeple-jack** [comment]
It builds Waybar v0.9.18, and I'm using the stable nixpkgs (23.05.3725.715d72e967ec)

My overlay is in system, maybe I should try it with home-manager instead.

I have Waybar set up to read from the standard .config files.

**Comment by zjeffer** [comment]
Yeah that's a pretty old version that probably doesn't have hyprland/workspaces yet.

**Comment by steeple-jack** [comment]
Is there a way to force nixpkgs to use a newer version or do I have to compile it myself from source? I'm new to NixOS sorry

**Comment by jficz** [comment]
in the overlay, you can specify a newer source to compile from

**Comment by steeple-jack** [comment]
After reading the wiki on Overlays, would something like this work?
```
nixpkgs.overlays = [
    (self: super: {
       waybar = super.waybar.overrideAttrs (oldAttrs: {
         src = self.fetchFromGithub {
           owner = "Alexays";
           repo = "waybar";
           rev = "e30fba0b8f875c7f35e3173be2b9f6f3ffe3641e";
           hash = "whatever the hash is here";
         };
         mesonFlags = oldAttrs.mesonFlags ++ [ "-Dexperimental=true" ];
       });
    })
  ];
```

**Comment by khaneliman** [comment]
> After reading the wiki on Overlays, would something like this work?
> 
> ```
> nixpkgs.overlays = [
>     (self: super: {
>        waybar = super.waybar.overrideAttrs (oldAttrs: {
>          src = super.fetchFromGithub {
>            owner = "Alexays";
>            repo = "waybar";
>            rev = "e30fba0b8f875c7f35e3173be2b9f6f3ffe3641e";
>            hash = "whatever the hash is here";
>          };
>          mesonFlags = oldAttrs.mesonFlags ++ [ "-Dexperimental=true" ];
>        });
>     })
>   ];
> ```

You could do that or use nixpkgs-wayland and reference the package from their if you dont want to maintain the overlay yourself. https://github.com/khaneliman/khanelinix/blob/main/modules/home/desktop/addons/waybar/default.nix#L130

**Comment by Vaisakhkm2625** [comment]
> > After reading the wiki on Overlays, would something like this work?
> > ```
> > nixpkgs.overlays = [
> >     (self: super: {
> >        waybar = super.waybar.overrideAttrs (oldAttrs: {
> >          src = super.fetchFromGithub {
> >            owner = "Alexays";
> >            repo = "waybar";
> >            rev = "e30fba0b8f875c7f35e3173be2b9f6f3ffe3641e";
> >            hash = "whatever the hash is here";
> >          };
> >          mesonFlags = oldAttrs.mesonFlags ++ [ "-Dexperimental=true" ];
> >        });
> >     })
> >   ];
> > ```
> 
> You could do that or use nixpkgs-wayland and reference the package from their if you dont want to maintain the overlay yourself. You dont even need the experimental = true since we call the hyprland dispatcher directly. https://github.com/khaneliman/khanelinix/blob/main/modules/home/desktop/addons/waybar/default.nix#L130
> 
> But, at this point, we should really have a new release for hyprland/workspaces since a lot of people just go off the latest stable and we've seen lots of people being surprised with Hyprland breaking wlr/workspaces and the latest release has very barebones hyprland/workspaces.

yes please... i am here new to nix, and struggle with configuring it with waybar and hyprland...

**Comment by Mythicsoul** [comment]
> Is there a way to force nixpkgs to use a newer version or do I have to compile it myself from source? I'm new to NixOS sorry

I'm in the same boat as you, but I grab the latest version from the unstable nix-channel.
```
  nixpkgs.overlays = let
    nixos-unstable = import <nixos-unstable> {};
  in [
    (final: prev: {
        inherit (nixos-unstable.pkgs) waybar;
    })
    (self: super: {
      waybar = super.waybar.overrideAttrs (oldAttrs: {
        mesonFlags = oldAttrs.mesonFlags ++ [ "-Dexperimental=true" ];
      });
    })
  ];
```
For this to work you need to add and then update the unstable nix-channel iirc.

**Comment by Alexays** [comment]
Closing — this was resolved/answered in the thread (config, support, or environment, not a Waybar bug still open). Feel free to reopen if you're still stuck. 🙏
